sqlite3.go

```diff
@@ -1487,296 +1487,294 @@ func (d *SQLiteDriver) Open(dsn string) (driver.Conn, error) {
-	//
-	// Workflow
-	//	- Authenticate
-	//		ON::SUCCESS		=> Continue
-	//		ON::SQLITE_AUTH => Return error and exit Open(...)
-	//
-	//  - Activate User Authentication
-	//		Check if the user wants to activate User Authentication.
-	//		If so then first create a temporary AuthConn to the database
-	//		This is possible because we are already successfully authenticated.
-	//
-	//	- Check if `sqlite_user`` table exists
-	//		YES				=> Add the provided user from DSN as Admin User and
-	//						   activate user authentication.
-	//		NO				=> Continue
-	//
-
 	// Create connection to SQLite
 	conn := &SQLiteConn{db: db, loc: loc, txlock: txlock}
 
-	// Password Cipher has to be registered before authentication
-	if len(authCrypt) > 0 {
-		switch strings.ToUpper(authCrypt) {
-		case "SHA1":
-			if err := conn.RegisterFunc("sqlite_crypt", CryptEncoderSHA1, true); err != nil {
-				return nil, fmt.Errorf("CryptEncoderSHA1: %s", err)
-			}
-		case "SSHA1":
-			if len(authSalt) == 0 {
-				return nil, fmt.Errorf("_auth_crypt=ssha1, requires _auth_salt")
-			}
-			if err := conn.RegisterFunc("sqlite_crypt", CryptEncoderSSHA1(authSalt), true); err != nil {
-				return nil, fmt.Errorf("CryptEncoderSSHA1: %s", err)
-			}
-		case "SHA256":
-			if err := conn.RegisterFunc("sqlite_crypt", CryptEncoderSHA256, true); err != nil {
-				return nil, fmt.Errorf("CryptEncoderSHA256: %s", err)
-			}
-		case "SSHA256":
-			if len(authSalt) == 0 {
-				return nil, fmt.Errorf("_auth_crypt=ssha256, requires _auth_salt")
-			}
-			if err := conn.RegisterFunc("sqlite_crypt", CryptEncoderSSHA256(authSalt), true); err != nil {
-				return nil, fmt.Errorf("CryptEncoderSSHA256: %s", err)
-			}
-		case "SHA384":
-			if err := conn.RegisterFunc("sqlite_crypt", CryptEncoderSHA384, true); err != nil {
-				return nil, fmt.Errorf("CryptEncoderSHA384: %s", err)
-			}
-		case "SSHA384":
-			if len(authSalt) == 0 {
-				return nil, fmt.Errorf("_auth_crypt=ssha384, requires _auth_salt")
-			}
-			if err := conn.RegisterFunc("sqlite_crypt", CryptEncoderSSHA384(authSalt), true); err != nil {
-				return nil, fmt.Errorf("CryptEncoderSSHA384: %s", err)
-			}
-		case "SHA512":
-			if err := conn.RegisterFunc("sqlite_crypt", CryptEncoderSHA512, true); err != nil {
-				return nil, fmt.Errorf("CryptEncoderSHA512: %s", err)
-			}
-		case "SSHA512":
-			if len(authSalt) == 0 {
-				return nil, fmt.Errorf("_auth_crypt=ssha512, requires _auth_salt")
-			}
-			if err := conn.RegisterFunc("sqlite_crypt", CryptEncoderSSHA512(authSalt), true); err != nil {
-				return nil, fmt.Errorf("CryptEncoderSSHA512: %s", err)
-			}
+	// All further database configuration *must* occur within this function, as
+	// it simplifies closing the database connection and unregistering associated
+	// functions or callbacks.
+	//
+	// Previously, this was handled at each return statement, which was error-prone
+	// and led to bugs because the database connection was not always closed.
+	err := func() error {
```

Comment:
Instead of this, we can use a named return value and a defer. That will also enable closing it on panic, which is another hole we have today.
```go
defer func() {
    if retErr != nil {
        conn.Close()
    }
}()
```

Reply:
Sorry, I was a little unclear. I just meant that using a `defer` would enable accounting for panics, not the named return itself.

The reason to close in the panic case is that this library cannot control what the caller of `Open` does. In particular, they could choose to swallow panics, and it would be good to at least avoid causing a memory leak in that situation.

Reply:
> Sorry, I was a little unclear. I just meant that using a defer would enable accounting for panics, not the named return itself.

Thanks and yeah was a little confused since in the given code snipped the defer would not close the db conn in the event of a panic (since the named return value would likely be nil). 

I updated the code to close the db connection if a panic occurs during setup.